File: android/app/src/main/java/com/connect/club/react/bridge/iosLike/ui/UserTogglesView.kt
package com.connect.club.react.bridge.iosLike.ui

import android.content.Context
import android.view.View
import androidx.core.view.isVisible
import com.connect.club.react.bridge.iosLike.Bool
import com.facebook.react.views.view.ReactViewGroup
import org.json.JSONObject

fun countIcons(video: Bool, audio: Bool): Int {
    var count = 0
    if (!video) count += 1
    if (!audio) count += 1
    return count
}

class UserTogglesView(context: Context?) : ReactViewGroup(context) {
    private var audioIconView: View? = null
    private var videoIconView: View? = null

    var audio = false
    var video = false

    fun onState(state: JSONObject) {
        //val isExpired = state.getBoolean("isExpired")
        val isReaction = state.getString("reaction") != "none"
        video = state.getBoolean("video")
        audio = state.getBoolean("audio")

        isVisible = (!video || !audio) && !isReaction// && !isExpired
        if (!isVisible) return
        requestLayout()

        videoIconView?.isVisible = !video
        audioIconView?.isVisible = !audio

        val count = countIcons(video, audio)

        val width = audioIconView!!.height
        if (audioIconView != null) {
            layout(left, top, left + width, top + width * count)
        }
    }

    fun showIfNeed() {
        isVisible = !video || !audio
    }

    override fun onLayout(changed: Boolean, left: Int, top: Int, right: Int, bottom: Int) {
        if ((!audio && video) || (audio && !video)) { audioIconView?.layout(0, 0, width, width) }
        if (!audio && !video) { videoIconView?.layout(0, 0, width, width) }
        if (!audio && !video) {
            videoIconView?.layout(0, 0, width, width)
            audioIconView?.layout(0, width, width, width * 2)
        }
    }


    override fun addView(child: View?, index: Int) {
        super.addView(child, index)

        if (child?.contentDescription == "audioIconView") {
            audioIconView = child
            child.isVisible = false
        }

        if (child?.contentDescription == "videoIconView") {
            videoIconView = child
            child.isVisible = false
        }
    }
}

File: ios/ReactNativeModules/ui/UserTogglesView.swift
//
//  UserTogglesView.swift
//  connectreactive
//
//  Created by Sergei Golishnikov on 13/05/2021.
//

import Foundation
import UIKit
import JitsiWebRTC

func countIcons(_ video: Bool, _ audio: Bool) -> CGFloat {
    var count: CGFloat = 0
    if !video { count += 1 }
    if !audio { count += 1 }
    return count
}

@objc(UserTogglesView)
class UserTogglesView: UIView {
    private weak var audioIconView: UIView?
    private weak var videoIconView: UIView?

    private var iconsCount: Int = 0
    
    var audio = false
    var video = false
    
    func onState(_ state: JSON) {
        let isReaction = state.reaction.string != "none"
        video = state.video.bool == true
        audio = state.audio.bool == true
        
        isVisible = (!video || !audio) && !isReaction
        if isHidden { return }
        
        videoIconView?.isHidden = video
        audioIconView?.isHidden = audio

        let count: Int = (video ? 0 : 1) + (audio ? 0 : 1)

        if let audioView = audioIconView, iconsCount != count {
            iconsCount = count
            frame.size.height = CGFloat(count) * audioView.frame.width
        }
        if !audio && video { audioIconView?.frame.origin.y = 0 }
        if !audio && !video { videoIconView?.frame.origin.y = 0 }
        if !audio && !video {
            videoIconView?.frame.origin.y = 0
            audioIconView?.frame.origin.y = frame.size.height / 2
        }
    }
    
    func showIfNeed() {
        isVisible = (!video || !audio)
    }
    
    override func addSubview(_ view: UIView) {
        super.addSubview(view)
        if view.accessibilityLabel == "audioIconView" {
            audioIconView = view
            view.isHidden = true
        }
        
        if view.accessibilityLabel == "videoIconView" {
            videoIconView = view
            view.isHidden = true
        }
    }
}
